smolyay/surrogate.py

```diff
@@ -156,6 +156,118 @@ def grid_generator(self, value):
         self._grid_generator = value
         self._reset_grid()
 
+    def derivative(self, x):
+        """Take the derivative of the surrogate at a given point.
+
+        The derivative of the surrogate can be obtained at a given
+        point inside the approximation domain after the surrogate
+        has been constructed and all of its coefficients have been computed.
+
+        Parameters
+        ----------
+        x: list (for dimension > 1), or float (for dimension = 1)
+            Input.
+
+        Returns
+        -------
+        list (for dimension > 1) or float (for dimension = 1)
+            Derivative of the surrogate in all dimensions.
+
+        Raises
+        ------
+        ValueError
+            For surrogate to be evaluated, function needs to be trained.
+            Input must lie in domain of surrogate.
+        IndexError
+            Input of the surrogate must be of length of the
+            function's dimensionality.
+        """
+        if self._coefficients is None:
+            raise ValueError('Function needs training!')
+        x = numpy.array(x, copy=False, ndmin=1)
+        if x.shape != (self.dimension, ):
+            raise IndexError('Input must match dimension of domain.')
+        if not (numpy.all(x >= self._domain[:, 0])
+                and numpy.all(x <= self._domain[:, 1])):
+            raise ValueError('x must lie in domain of surrogate')
+
+        gradient = []
+        # transform point into basis domain
+        x_scaled = self._mapdomain(x, self._domain, [[-1, 1]]*self.dimension)
+        # evaluate the gradient
+        for ni in range(self.dimension):
+            value = 0
+            for coeff, basis in zip(self.coefficients,
+                                    self.grid.basis_functions):
+                if self.dimension > 1:
+                    term = numpy.prod(
+                        [f.derivative(x) if dim == ni else f(x) for x, f, dim
+                         in zip(x_scaled, basis, range(self.dimension))])
+                else:
+                    term = coeff*basis.derivative(x_scaled[0])
```

Comment:
I think no factor of `coeff` here because it is on the next line?
```suggestion
                    term = basis.derivative(x_scaled[0])
```